dev/com.ibm.ws.sipcontainer/src/com/ibm/ws/sip/stack/transport/sip/netty/SipConnLink.java

```diff
@@ -0,0 +1,523 @@
+				        }
+						m_outMessages.addLast((MessageContext) messageContext);
+					}
+				}
+			}
+		}
+		catch (IOException e) {
+			connectionError(e);
+		}
+		
+		if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+            Tr.debug(tc,  "write",  "exit [" + System.identityHashCode(messageContext) + ']');
+        }
+	}
+
+	/**
+	 * sends out one message
+	 * 
+	 * @param buffer
+	 *            message to be sent
+	 * @return true if completed immediately
+	 */
+	private boolean sendNow(MessageContext messageContext) throws IOException {
+		if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+            Tr.debug(tc, "sendNow", "entry [" + System.identityHashCode(messageContext) + ']');
+        }
+
+		if (PerformanceMgr.getInstance().isTaskDurationOutboundQueuePMIEnabled() && messageContext != null) {
+			// Message is out of the queue - measure how much time it was queued
+			if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+	            Tr.debug(tc, "sendNow", "measure task duration");
+	        }
+			
+			PerformanceMgr.getInstance().measureTaskDurationOutboundQueue(messageContext.getSipContainerQueueDuration().takeTimeMeasurement());
+		}
+		
+		if(messageContext !=null) {
+			if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+				Tr.debug(tc,  "sendNow", "update QueueMonitoring outbound queue statistics - task dequeued");
+			}
+			
+			PerformanceMgr.getInstance().updateQueueMonitoringTaskDequeuedFromOutboundQueue();
+		}
+		
+		SipMessageByteBuffer sipBuffer = messageContext.getSipMessageByteBuffer();
+		messageContext.setSipMessageByteBuffer(null);
+
+		ByteBuf buffer = stackBufferToByteBuf(sipBuffer);
+		final ChannelFuture writeFuture = m_channel.writeAndFlush(buffer);	
+
+		if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+			Tr.debug(tc,  "sendNow", "exit [" + System.identityHashCode(messageContext) + ']');
+		}
+		
+		boolean complete = writeFuture.isDone();
+		
+		if (!complete) {
+			writeFuture.addListener(messageContext);
+		}
+		
+		return complete;
+	}
+
+	/**
+	 * @see com.ibm.ws.sip.stack.transaction.transport.connections.SIPConnection#getMessageParser()
+	 */
+	public MessageParser getMessageParser() {
+		return m_messageParser;
+	}
+
+	// -----------------------------
+	// ConnectionLink implementation
+	// -----------------------------
+
+
+	/**
+	 * @see com.ibm.wsspi.channelfw.ConnectionReadyCallback#destroy(java.lang.Exception)
+	 */
+	public void destroy(Exception e) {
+		connectionError(e);
+	}
+
+	/*
+	 * (non-Javadoc)
+	 * 
+	 * @see com.ibm.ws.sip.stack.transaction.transport.connections.SIPConnectionAdapter#writeComplete()
+	 */
+	public void writeComplete(MessageContext messageContext) {
+		if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+			Tr.debug(tc,  "writeComplete", "entry [" + System.identityHashCode(messageContext) + ']');
+		}
+
+		// TODO should we release a ByteBuf created by Unpooled.buffer
```

Comment:
TODO

Reply:
This 2do we might need to look at.  I'll consider it covered by design issue 21039.   I'll delete the block of comment code below for now.